astropy/wcs/src/wcslib_prjprm_wrap.c

```diff
@@ -0,0 +1,1025 @@
+        self->x->flag = 0;
+        if (self->owner) ((PyCelprm*)self->owner)->x->flag = 0;
+        Py_RETURN_NONE;
+    }
+
+    if (PyFloat_Check(value) || PyLong_Check(value)) {
+        data = PyFloat_AsDouble(value);
+        if (data == -1.0 && PyErr_Occurred()) {
+            return NULL;
+        }
+
+    } else if (PyUnicode_Check(value)) {
+        flt_value = PyFloat_FromString(value);
+        if (!flt_value) return NULL;
+        data = PyFloat_AsDouble(flt_value);
+        Py_DECREF(flt_value);
+        if (data == -1.0 && PyErr_Occurred()) {
+            return NULL;
+        }
+
+    } else {
+        if (PyArray_Converter(value, &value_array) == NPY_FAIL) {
+            return NULL;
+        }
+
+        size = PyArray_SIZE(value_array);
+        if (size != 1) {
+            Py_DECREF(value_array);
+            PyErr_SetString(PyExc_ValueError,
+                "PV value must be a scalar-like object or None.");
+            return NULL;
+        }
+
+        scalar = PyArray_ToScalar(PyArray_DATA(value_array), value_array);
+        Py_DECREF(value_array);
+        if (!scalar) {
+            Py_DECREF(scalar);
+            PyErr_SetString(PyExc_TypeError, "Unable to convert value to scalar.");
+        }
+
+        PyArray_CastScalarToCtype(scalar, &data, dbl_descr);
+        Py_DECREF(scalar);
+        if (PyErr_Occurred()) {
+            return NULL;
+        }
+    }
+
+    data = (isnan(data)) ? UNDEFINED : data;
+
+    if (!is_dbl_equal(self->x->pv[idx], data)) {
+        self->x->flag = 0;
+        if (self->owner) ((PyCelprm*)self->owner)->x->flag = 0;
+    }
+    self->x->pv[idx] = data;
+
+    Py_RETURN_NONE;
+}
+
+
+static PyObject* PyPrjprm_get_bounds(PyPrjprm* self, void* closure)
+{
+    if (is_prj_null(self)) {
+        return NULL;
+    } else {
+        return get_int("bounds", self->x->bounds);
+    }
+}
+
+
+static int PyPrjprm_set_bounds(PyPrjprm* self, PyObject* value, void* closure)
+{
+    if (is_prj_null(self) || is_readonly(self)) {
+        return -1;
+    } else if (value == Py_None) {
+        self->x->bounds = 0;
+        return 0;
+    } else {
+        return set_int("bounds", value, &self->x->bounds);
+    }
+}
+
+
+static PyObject* PyPrjprm_get_w(PyPrjprm* self, void* closure)
+{
+    Py_ssize_t size = 10;
+    int k;
+    double *w;
+    PyObject* w_array;
+
+    if (is_prj_null(self)) {
+        return NULL;
+    }
+
+    w = malloc(size * sizeof(double));
+    if (w == NULL) {
+        PyErr_SetString(PyExc_MemoryError,
+                        "Could not allocate memory for 'w' array.");
+        return NULL;
+    }
+
+    for (k = 0; k < size; k++) {
+        if (self->x->w[k] == UNDEFINED) {
+            w[k] = (double) NPY_NAN;
+        } else {
+            w[k] = self->x->w[k];
+        }
+    }
+
+    w_array = (PyObject*) PyArray_SimpleNewFromData(1, &size, NPY_DOUBLE, w);
+    if (w_array == NULL) return NULL;
+    PyArray_ENABLEFLAGS(w_array, NPY_ARRAY_OWNDATA);
```

Comment:
I'm not very familiar with this, my understanding is that this isn't the prescribed way to avoid a memory leak:

https://numpy.org/devdocs/reference/c-api/array.html#c.PyArray_SetBaseObject

This statement is unequivocal:

> If you construct an array by passing in your own memory buffer as a parameter, you need to set the array’s base property to ensure the lifetime of the memory buffer is appropriate.

Reply:
> I'm not very familiar with this, ...
me neither...  but this was taken directly from code already in `astropy`:

https://github.com/astropy/astropy/blob/295b2d0e42639b919b9217a70338ae24e29c4f7e/astropy/io/fits/src/compressionmodule.c#L1063-L1071

I just rewrote this using `PyArray_SimpleNew()`.